src/array-map.ts

```diff
@@ -0,0 +1,293 @@
+
+/**
+ * @private
+ */
+export class TrackedArrayMap<Element = unknown, MappedTo = unknown>
+  implements MappedArray<Element[], MappedTo>
+{
+  // Tells TS that we can array-index-access
+  [index: number]: MappedTo;
+
+  // these can't be real private fields
+  // until @cached is a real decorator
+  private _mapCache = new WeakMap<Element & object, MappedTo>();
+  private _dataFn: () => readonly Element[];
+  private _mapper: (element: Element) => MappedTo;
+
+  constructor(
+    data: () => readonly Element[],
+    map: (element: Element) => MappedTo,
+  ) {
+    this._dataFn = data;
+    this._mapper = map;
+
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
+
+    /**
+     * This is what allows square-bracket index-access to work.
+     *
+     * Unfortunately this means the returned value is
+     * Proxy -> Proxy -> wrapper object -> *then* the class instance
+     *
+     * Maybe JS has a way to implement array-index access, but I don't know how
+     */
+    return new Proxy(this, {
+      get(_target, property) {
+        if (typeof property === "string") {
+          let parsed = parseInt(property, 10);
+
+          if (!isNaN(parsed)) {
+            return self[AT](parsed);
+          }
+        }
+
+        return self[property as keyof MappedArray<Element[], MappedTo>];
+      },
+      // Is there a way to do this without lying to TypeScript?
+    }) as TrackedArrayMap<Element, MappedTo>;
+  }
+
+  @signal
+  get _records(): (Element & object)[] {
+    let data = this._dataFn();
+
+    if (!data.every((datum) => typeof datum === "object")) {
+      throw new Error(
+        `Every entry in the data passed to \`map\` must be an object.`,
+      );
+    }
+
+    return data as Array<Element & object>;
+  }
+
+  values = () => [...this];
+
+  get length() {
+    return this._records.length;
+  }
+
+  [Symbol.iterator](): Iterator<MappedTo> {
+    let i = 0;
+
+    return {
+      next: () => {
+        if (i >= this.length) {
+          return { done: true, value: null };
+        }
+
+        let value = this[AT](i);
+
+        i++;
+
+        return {
+          value,
+          done: false,
+        };
+      },
+    };
+  }
+
+  /**
+   * @private
+   *
+   * don't conflict with
+   *   https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at
+   */
+  [AT] = (i: number) => {
+    let record = this._records[i];
+
+    if (!record) {
+      throw new Error(
+        `Expected record to exist at index ${i}, but it did not. ` +
+          `The array item is expected to exist, because the map utility resource lazily iterates along the indices of the original array passed as data. ` +
+          `This error could happen if the data array passed to map has been mutated while iterating. ` +
+          `To resolve this error, do not mutate arrays while iteration occurs.`,
+      );
+    }
+
+    let value = this._mapCache.get(record);
```

Comment:
```suggestion
    const value = this._mapCache.get(record);
```

Reply:
let's leave this to linting, no strong opinion here, but linting-style comments in PRs are a bit noisy <3